Make order setting mandatory for Realm config

ywangd opened the pull request:
The order parameter must be explicitly specified for each realm.
This is a breaking change and will start take effect in 8.0

Resolves: #37614

ywangd committed:
WIP: making realm order config mandatory

ywangd committed:
Support explicit order parameter for RealmConfig

Reserved realm and default native realms now use the new constructor to
specifiy order explicitly

ywangd committed:
Fix security plugin tests for required order param

ywangd commented:
@tvernum This PR is still WIP. But I'd like to get a PR in early its easier for you to review and set me on the right track. 

Current changes are focused on making the order parameter mandatory and fixing relevant tests. Others things remaining are:
* No two realm should have the same order
* Warning in 7.x
* Docs update
* Breaking change update

ywangd committed:
Fix more tests for required order param

ywangd committed:
Fix ci/2 failures

ywangd commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into issue-37614-realm-order

ywangd committed:
Fix more require order parameter failure

ywangd commented:
@elasticmachine run elasticsearch-ci/2

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmConfig.java:
```diff
@@ -24,11 +24,24 @@
+            );
+        } else {
+            this.order = getSetting(RealmSettings.ORDER_SETTING);
+        }
```
I would prefer the `order` be a primitive, and we resolve the order from the settings in the original constructor.

Behaviours around _if this parameter is null, do this other thing_ can end up getting quite confusing in a large code base.

ywangd replied:
I agree `null` is a hack. But there is no easy way to tell whether a primitive value is indeed missing.

ywangd replied:
Changed to `int`.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmConfig.java:
```diff
@@ -24,11 +24,24 @@
+        this(identifier, settings, env, threadContext, null);
+    }
+
+    public RealmConfig(RealmIdentifier identifier, Settings settings, Environment env, ThreadContext threadContext, Integer order) {
```
I think it would be better for this constructor to be  protected, and update the tests to pass through real `Settings` object with an order config.

I know that's going to be a pain - so please feel free to convince me if you feel otherwise.

ywangd replied:
Are you suggesting the caller should create a `Settings` object that actually includes an `order`? In this case, this constructor is no longer necessary? Am I getting this right?

ywangd replied:
I initially actual prefer to **not** adding another constructor. But the number of call site changes made me change my mind. But I am happy to drop the constructor and make more call site changes if you favour this approach. Please let me know. Thanks

ywangd replied:
Changed to `int`.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmConfig.java:
```diff
@@ -24,11 +24,24 @@
+        } else if (order == null && hasSetting(RealmSettings.ORDER_SETTING.apply(type())) == false) {
+            throw new IllegalArgumentException("'order' is a mandatory parameter for realm config. " +
+                "Found invalid realm config: '" + identifier.name + "'\n" +
+                "Please see the breaking changes documentation."
```
Do we have a test for this validation (e.g. `RealmSettingsTests`) ?
I can't see one, but given the number of tests changes, I may have missed it.

ywangd replied:
I haven't added the tests yet mainly because I was not sure how the change will turn out (e.g. the above discussion may result in implementation change). I'll add them once I am certain about the changes.

ywangd replied:
Tests added.

ywangd commented:
@tvernum I plan to go ahead with following changes. Please let me know if this works for you:
* Drop the new constructor. The only constructor will mandate that the `Settings` object has an `order` setting.
* Change relevant call sites in tests to always add `order` into `Settings`.
Personally I think above approach works better since it does not change production code to cater tests.

To ensure no two realms having the same order, it seems that some logic needs to be added in `Security.java`. But shall we enforce this requirement to builtin realms (file, index) that are added automatically? Currently they are both added with default order of `Integer.MAX_VALUE`. It does not seem to be very meaningful to have different order values for them.

ywangd commented:
Just realised that the `Settings` object is also referenced from the `Environment` object. The following code snippet shows how we add default file realm:
```java
realms.add(fileRealm.create(new RealmConfig(
                    new RealmConfig.RealmIdentifier(FileRealmSettings.TYPE, "default_" + FileRealmSettings.TYPE),
                    settings, env, threadContext)));
```

To add the `order` setting, above code would change to something like the followings:
```java
var realmIdentifier =
                new RealmConfig.RealmIdentifier(FileRealmSettings.TYPE, "default_" + FileRealmSettings.TYPE);
            realms.add(fileRealm.create(new RealmConfig(
                realmIdentifier,
                Settings.builder().put(settings)
                    .put(RealmSettings.realmSettingPrefix(realmIdentifier) + "order", Integer.MIN_VALUE).build(),
                env, threadContext)));
```

Note that a new `Settings` object is created with the addition of `order` setting. But the `env` object still references old `Settings` object. Will this create a problem somewhere else in the code path?

ywangd committed:
Enforce no same order for realms.

Add tests for mandatory order setting of RealmConfig.
Fix broken tests due to unique order setting requirement.

ywangd commented:
@tvernum I would like to change my view on this again ... Sorry for the back and forth. I kept the new constructor but changed the `order` parameter to be of a primitive int type. The Constructor remains public as it is used by `Realms` and `ReservedRealm`. 

A few reasons to go with this approach:
* The mandatory `order` setting is not 100% strict, because native realms can be added without any settings. So currently it feels we have an escape hatch for this rule, which seems to make sense for another constructor.
* When default native realms are added, they don't alter the existing `Settings` object. I'd like to follow this pattern.
* The `Settings` object is referenced in multiple places, `Realms`, `Environment`, every realm and likely more. I am unsure what would be the consequence if they each reference a different Settings object (mostly the same content, but distinct object). 
* This approach is likely more extensible if we ever wanna make another setting mandatory or some other requirement, e.g. it could refactor to replace the `int` parameter with a value provider.

Other changes I have pushed are:
* [Test file for RealmConfig](https://github.com/elastic/elasticsearch/pull/51195/files#diff-5663daa5e8b412da8ec1d00c35f8f7be) - Please let me know whether they are idomatic in elasticsearch's term.
* [Enforce unique order for each realm in Realms](https://github.com/elastic/elasticsearch/pull/51195/files#diff-ff593e0560017b528a889b09ee2de166) - Note this is only applied to user configure realms. I feel this does its job without worring about orders of default native realms (Integer.MIN_VALUE). 
* [Updated tests for Realms](https://github.com/elastic/elasticsearch/pull/51195/files#diff-53d7973edbfba7ef205c1d7d1b32b5a3)

ywangd committed:
Fix duplicate order for tests

ywangd committed:
Add realm order to breaking change doc

ywangd committed:
Start updating docs for realm order

ywangd committed:
Add missing order config in docs

Also update descriptions

ywangd commented:
Tried to identify all places in docs that are relevant to the `order` setting. To my surprise, **not** many places need update. Here is a list of files that I have changed for review convenience:
* [security.asciidoc](https://github.com/elastic/elasticsearch/pull/51195/files#diff-fd39b6fcbbb75d4e4d0b52c1a9d2078e)
* [security-settings.asciidoc](https://github.com/elastic/elasticsearch/pull/51195/files#diff-f477b3b9a9b90c0b5b1814dfc5570ce7)
* [configuring-ldap-realm.asciidoc](https://github.com/elastic/elasticsearch/pull/51195/files#diff-2fe19257a3c1352697639a03dd2da03a)
* [configuring-pki-realm.asciidoc](https://github.com/elastic/elasticsearch/pull/51195/files#diff-d58864db3ef48edb02328873c9ec1556)

ywangd commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into issue-37614-realm-order

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

lcawl commented on docs/reference/migration/migrate_8_0/security.asciidoc:
```diff
@@ -6,6 +6,29 @@
+==== The realm `order` setting is required
+
+The `xpack.security.authc.realms.*.*.order` setting is now required and must be
+specified for each explicitly configured realms. Their values must be unique.
```
```suggestion
specified for each explicitly configured realm. Each value must be unique.
```

ywangd committed:
Update docs/reference/migration/migrate_8_0/security.asciidoc

Co-Authored-By: Lisa Cawley <lcawley@elastic.co>

lcawl commented on x-pack/docs/en/security/authentication/configuring-ldap-realm.asciidoc:
```diff
@@ -206,6 +206,7 @@ xpack:
       realms:
         ldap:
           ldap1:
+            order: 0
```
There's another line in this file that I think should be changed: 

> If you are configuring multiple realms, you should also explicitly set the `order` attribute ....

I think it should be changed to "you must also explicitly set the `order` attribute".

ywangd replied:
Good catch. Will update. Thanks

ywangd commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into issue-37614-realm-order

ywangd committed:
Update for docs feedback

lcawl commented on x-pack/docs/en/security/authentication/configuring-pki-realm.asciidoc:
```diff
@@ -61,6 +61,7 @@ xpack:
       realms:
         pki:
           pki1:
+            order: 1
```
This page also has the following:

> If you are configuring multiple realms, you should explicitly set the order attribute.

... which I think should be changed to "you must explicitly set the order attribute..."

ywangd replied:
Thanks. It's updated now. I also searched through the documentations to make sure there is no similar situation like this anymore.

ywangd replied:
Here are the similar changes that are made based on above search result

ywangd replied:
https://github.com/elastic/elasticsearch/pull/51195/commits/994c0af8d053c683eaca306770beaf3207c16dc5

ywangd committed:
Update doc to address feedback

ywangd committed:
More wording changes based on feedback

albertzaharovits reviewed (commented):
I've done a first pass.
I appreciate the effort involved, but it's one of those cases where I would've made a different choice.

But I think this works and I'll LGTM when I get to look a second time.

albertzaharovits commented on docs/reference/migration/migrate_8_0/security.asciidoc:
```diff
@@ -6,6 +6,29 @@
+For example, the following configuration is invalid:
+[source,yaml]
+--------------------------------------------------
+xpack.security.authc.relams.kerberos.kerb1:
```
this is duplicated, I'll remove, it was from a review started in a different tab.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmConfig.java:
```diff
@@ -27,9 +27,24 @@ public RealmConfig(RealmIdentifier identifier, Settings settings, Environment en
+        this.env = env;
         this.threadContext = threadContext;
+        this.enabled = getSetting(RealmSettings.ENABLED_SETTING);
+        this.order = order;
```
The new constructor is mostly used by test code, but it is also used by production code, see [Realms](https://github.com/elastic/elasticsearch/pull/51195/files#diff-ff593e0560017b528a889b09ee2de166R315) and [ReservedRealm](https://github.com/elastic/elasticsearch/pull/51195/files#diff-13b2f866e83efa400fc0999b3a6834a6L66).

I also think adding a new constructor is not ideal due to some of the concerns you mention. It however seems to be a better choice comparing to the other choice, i.e. Add `order` into the `settings` parameter on caller side so that it can just keep using the existing constructor. 

This approach should work fine for tests. But in production code, the `settings` object is referenced in many other classes, e.g. `Environment`, `Realms`, etc. So when adding FileRealm (in [Realms.java](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java#L309)), if we modify the `settings` object to add `order`, we will need to do something like the follows:

```java
var newSettings = Settings.Builder().put(settings).build();
... new RealmConfig(..., newSetings, ...), ...;
```
So the `newSettings` is now passed into RealmConfig which will store a reference to it. Now we have two `Settings` object across the application. They are mostly identical, only differs in the `order` parameter. And this is just FileRealm, the same applies to native realm, for which we will have a third `settings`. I personally think this is worse than adding a new constructor. That is why the new constructor is added so we can avoid manipulating the settings object.

With above being said, I realise it is making an assumption that the logic for mandating the `order` setting resides in the `RealmConfig` class. If we can move away from this assumption, like you suggested, enforce the logic in `Realms.initRealms`, we can probably make the change set much smaller. 

Initially I thought about adding the logic into `Security` but that was too *high* level. So we decided to add it in `RealmConfig` itself, where it is *low* level the logic is close to its natural owner. But maybe we need something in between. `Realms` could be the right candidate especially it is where the order uniqueness is enforced.

albertzaharovits replied:
My choice would also be to only inspect the `order` setting in`Realms#initRealms` (where the `order` value duplication is checked). That's because the `order` is only really used in that place, specifying the **initial** order in the realm chain (which would then change as we reorder the chain during authentications), so it's not suitable as a final local variable.
This way we would also avoid having to specify the `order` parameter for the "default" realms (`file` and `native`) which is simply ignored anyways.

Your approach works as well and I can't argue strongly against it, so this LGTM to me even though I would've approached it differently.

ywangd commented on docs/reference/migration/migrate_8_0/security.asciidoc:
```diff
@@ -6,6 +6,29 @@
+For example, the following configuration is invalid:
+[source,yaml]
+--------------------------------------------------
+xpack.security.authc.relams.kerberos.kerb1:
```
Good catch thanks!

ywangd commented on x-pack/docs/en/security/authentication/configuring-active-directory-realm.asciidoc:
```diff
@@ -4,7 +4,7 @@ realm and map Active Directory users and groups to roles in the role mapping fil
+If you are configuring multiple realms, you must also 
 explicitly set the `order` attribute to control the order in which the realms 
 are consulted during authentication. 
 +
```
Makes sense will do

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmConfig.java:
```diff
@@ -27,9 +27,24 @@ public RealmConfig(RealmIdentifier identifier, Settings settings, Environment en
         this.env = env;
+        this.threadContext = threadContext;
         this.enabled = getSetting(RealmSettings.ENABLED_SETTING);
+        if (hasSetting(RealmSettings.ORDER_SETTING.apply(type())) == false) {
```
I am also tempted to move the logic into `Reamls#initRealms` (see below) comments for details. But your examples show that by itself is *not* sufficient. Both of the two examples are *not* covered if the logic is just in `Realms#initRealms`, while they do get covered if the logic is in `RealmConfig`. I am now again unsure about what would be the better approach.

albertzaharovits replied:
Maybe I've not made myself clear
> will throw an exception in InternalRealms.getBootstrapChecks so the stacktrace for it is funky.

I mean that there is a minor inconvenience that the validation that the `order` parameter is defined is performed inside a different code than the code that deals with realm construction.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:
```diff
@@ -218,9 +219,13 @@ public Realm realm(String name) {
             nameToRealmIdentifier.computeIfAbsent(realm.name(), k ->
                 new HashSet<>()).add(RealmSettings.realmSettingPrefix(realm.type()) + realm.name());
+            orderToRealmIdentifier.computeIfAbsent(realm.order(), k -> new HashSet<>())
+                .add(RealmSettings.realmSettingPrefix(realm.type()) + realm.name());
```
Good point. I'll make them consistent. Probably always use the setting key which is more explicit.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/test/SecuritySettingsSource.java:
```diff
@@ -141,8 +141,8 @@ public Settings nodeSettings(int nodeOrdinal) {
-                .put("xpack.security.authc.realms." + NativeRealmSettings.TYPE + ".index.order", "1")
+                .put("xpack.security.authc.realms." + FileRealmSettings.TYPE + ".file.order", Integer.MIN_VALUE)
+                .put("xpack.security.authc.realms." + NativeRealmSettings.TYPE + ".index.order", String.valueOf(Integer.MIN_VALUE + 1))
                 .put("xpack.license.self_generated.type", "trial");
```
You are right. I think I edited the wrong file. Nevermind, I'll get rid of them. Thanks

ywangd committed:
Address feedback for docs

ywangd committed:
Address feedback to revert accident change

ywangd committed:
Merge remote-tracking branch 'origin/master' into issue-37614-realm-order

ywangd committed:
Update based on discussion with Tim.

ywangd committed:
Address feedback for consistent err msg

ywangd commented:
Following discussion with Tim, I have made the following changes:

* *Removed* the new constructor of RealmConfig
* Still kept the logic of checking mandatory `order` setting in RealmConfig. I think this class is still overall the better place for enforcing the check because it is the choke point for many different upstream use cases, including some command line tools. So we will have a consistent report on the missing order setting no matter what the entry point is.
* Since RealmConfig no longer has a special constructor for handling missing order setting, it is now the caller's responsibility to ensure this setting exists for reserved, file and native realms (when they are added automatically). I am aware that this approach creates new settings object along the way. This is however an existing issue in other places. A separate issue (#51387) has been created to tackle it. 
* The approach to ensure uniqueness of realm orders remains the same (i.e. in `Realms#initRealms`)
* Again most of the changes are for test classes (making sure each of them has an order setting configured).

albertzaharovits commented on x-pack/docs/en/security/authentication/configuring-ldap-realm.asciidoc:
```diff
@@ -21,10 +21,9 @@ However, multiple bind operations might be needed to find the correct user DN.
+the `url` and `order` of the LDAP server, and set `user_search.base_dn` to the
+container DN where the users are searched for.
+The `order` attribute to control the order in which the realms are consulted during 
 authentication. See <<ref-ldap-settings>> for all of the options you can set for 
```
This last phrase doesn't make sense, I would avoid describing again what `order` does.

ywangd replied:
Will do thanks

albertzaharovits commented on x-pack/docs/en/security/authentication/configuring-ldap-realm.asciidoc:
```diff
@@ -73,7 +72,7 @@ realms you specify are used for authentication. If you also want to use the
-`user_dn_templates` option. If you are configuring multiple realms, you should
+`user_dn_templates` option. If you are configuring multiple realms, you must
 also explicitly set the `order` attribute to control the order in which the
 realms are consulted during authentication.
```
I would remove
> If you are configuring multiple realms, you must ...

altogether.

ywangd replied:
Somehow missed this occurrence ... Will remove. Thanks

albertzaharovits commented on x-pack/docs/en/security/authentication/custom-realm.asciidoc:
```diff
@@ -89,11 +89,10 @@ under the `xpack.security.authc.realms` namespace.
+At a minimum, you must explicitly set the `order` attribute to control the
+order in which the realms are consulted during authentication. You must also
+make sure each configured realm has a distinct `order` setting. In the event
+that two or more realms have the same `order`, the cluster will fail to start.
```
```suggestion
that two or more realms have the same `order`, the node will fail to start.
```

albertzaharovits commented on x-pack/docs/en/security/authentication/realm-chains.asciidoc:
```diff
@@ -5,9 +5,9 @@
+consulted first). You must make sure each configured realm has a distinct
 `order` setting. In the event that two or more realms have the same `order`,
-they are processed in `name` order.
+the cluster will fail to start.
```
```suggestion
the node will fail to start.
```

albertzaharovits reviewed (approved):
LGTM, thank you Yang!

ywangd committed:
Update x-pack/docs/en/security/authentication/custom-realm.asciidoc

Co-Authored-By: Albert Zaharovits <albert.zaharovits@gmail.com>

ywangd committed:
Update x-pack/docs/en/security/authentication/realm-chains.asciidoc

Co-Authored-By: Albert Zaharovits <albert.zaharovits@gmail.com>

ywangd committed:
Address feedback for docs

ywangd committed:
Merge remote-tracking branch 'origin/master' into issue-37614-realm-order

albertzaharovits commented:
@ywangd I just remembered, we need to follow-up this PR with a another one that adds a deprecation check in https://github.com/elastic/elasticsearch/blob/7.x/x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/NodeDeprecationChecks.java . This should inform users that realm definition sections without the `order` parameter are invalid. The PR should be based against the `7.x` branch (these checks do not exist on `master`).

ywangd commented:
@albertzaharovits Warning/Deprecation message for 7.x is on my to-do list. Thanks for pointing me to the right file to start with.

tvernum commented on docs/reference/migration/migrate_8_0/security.asciidoc:
```diff
@@ -6,6 +6,29 @@
+[float]
+==== The realm `order` setting is required
+
+The `xpack.security.authc.realms.*.*.order` setting is now required and must be
```
```suggestion
The `xpack.security.authc.realms.{type}.{name}.order` setting is now required and must be
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:
```diff
@@ -305,15 +310,34 @@ public void usageStats(ActionListener<Map<String, Object>> listener) {
+            .map(entry -> entry.getKey() + ": " + entry.getValue())
+            .collect(Collectors.joining("; "));
+        if (Strings.hasText(duplicateOrders)) {
+            throw new IllegalArgumentException("Found multiple realms configured with the same order: " + duplicateOrders + "");
```
```suggestion
            throw new IllegalArgumentException("Found multiple realms configured with the same order: " + duplicateOrders);
```

tvernum reviewed (approved):
LGTM

ywangd committed:
Update docs/reference/migration/migrate_8_0/security.asciidoc

Co-Authored-By: Tim Vernum <tim@adjective.org>

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java

Co-Authored-By: Tim Vernum <tim@adjective.org>

ywangd committed:
Merge remote-tracking branch 'origin/master' into issue-37614-realm-order
